src/renderers/shared/reconciler/ReactCompositeComponent.js

```diff
@@ -808,8 +808,10 @@ var ReactCompositeComponentMixin = {
   attachRef: function(ref, component) {
     var inst = this.getPublicInstance();
     invariant(inst != null, 'Stateless function components cannot have refs.');
+    var componentInstance = component.getPublicInstance();
```

Comment:
semantics, but let's call this `publicInstance` or `publicComponentInstance` instead of `componentInstance`.  The problem is that the `component` variable passed in is technically also a componentInstance (just, it's the internal one).